src/machine/machine.c

```diff
@@ -523,6 +524,36 @@ int machine_finalize(Machine *m) {
         return 0;
 }
 
+static bool machine_validate_unit(Machine *m) {
+        int r;
+        _cleanup_free_ char *unit = NULL;
+        _cleanup_free_ char *cgroup = NULL;
+
+        r = cg_pid_get_unit(m->leader, &unit);
+        if (!r && streq(m->unit, unit))
+                return true;
+
+        if (r == -ESRCH)
+                log_info("Machine unit is inactive, but the leader process is exited. "
+                        "machine: %s, leader: "PID_FMT", unit: %s.", m->name, m->leader, m->unit);
+        else if (r)
+                log_info_errno(r, "Can not get unit from cgroup. "
+                        "machine: %s, leader: "PID_FMT", unit: %s, error: %m", m->name, m->leader, m->unit);
+        else if (unit && !streq(m->unit, unit))
+                log_info("Machine unit name not match. "
+                        "machine: %s, leader: "PID_FMT", machine unit: %s, real unit: %s", m->name, m->leader, m->unit, unit);
```

Comment:
Shouldn't this immediately return `false` after any of those warnings are printed?

Reply:
Indeed.